File: Anul II/Medii de proiectare si programare/Laborator/lab08/Java/networking/src/main/java/com/ubb/protocol/ServicesProxy.java
package com.ubb.protocol;

import com.ubb.IMainObserver;
import com.ubb.IContestServices;
import com.ubb.exceptions.ContestDataException;
import com.ubb.model.*;
import com.ubb.dto.RaceDTO;
import com.ubb.protocol.request.*;
import com.ubb.protocol.response.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ServicesProxy implements IContestServices {
    private final String host;
    private final int port;
    private IMainObserver client;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;

    private final BlockingQueue<Response> responseQueue;
    private volatile boolean finished;

    public ServicesProxy(String host, int port) {
        this.host = host;
        this.port = port;
        responseQueue = new LinkedBlockingQueue<>();
    }

    @Override
    public User login(String username, String password, IMainObserver client) throws ContestDataException {
        initializeConnection();
        sendRequest(new LoginRequest(username, password));
        Response response = readResponse();
        if (response instanceof OKResponse) {
            this.client = client;
            return ((OKResponse) response).getUser();
        }
        if (response instanceof ErrorResponse) {
            closeConnection();
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        return null;
    }

    @Override
    public void logout(User user, IMainObserver client) throws ContestDataException {
        sendRequest(new LogoutRequest(user));
        Response response = readResponse();
        closeConnection();
        if (response instanceof ErrorResponse){
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
    }

    @Override
    public Collection<Participant> getParticipantsByTeam(Long teamID) throws ContestDataException {
        sendRequest(new GetParticipantsByTeamRequest(teamID));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        GetParticipantsByTeamResponse participantsResponse = (GetParticipantsByTeamResponse) response;
        return participantsResponse.getParticipants();
    }

    @Override
    public Optional<Participant> getParticipantByData(Participant participant) throws ContestDataException {
        sendRequest(new GetParticipantByDataRequest(participant));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        ParticipantResponse participantResponse = (ParticipantResponse) response;
        return Optional.of(participantResponse.getParticipant());
    }

    @Override
    public Optional<Race> getRaceByName(String name) throws ContestDataException {
        sendRequest(new GetRaceByNameRequest(name));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        RaceByNameResponse raceResponse = (RaceByNameResponse) response;
        return Optional.of(raceResponse.getRace());
    }

    @Override
    public Collection<RaceEntry> getEntriesByRace(Long raceID) throws ContestDataException {
        sendRequest(new GetEntriesByRaceRequest(raceID));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        EntriesByRaceResponse entriesResponse = (EntriesByRaceResponse) response;
        return entriesResponse.getEntries();
    }

    @Override
    public Collection<Race> getRacesByEngineCapacity(Integer engineCapacity) throws ContestDataException {
        sendRequest(new GetRacesByEngineCapacityRequest(engineCapacity));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        RacesByCapacityResponse racesResponse = (RacesByCapacityResponse) response;
        return racesResponse.getRaces();
    }

    @Override
    public Collection<RaceDTO> getRacesWithParticipantCount() throws ContestDataException {
        sendRequest(new GetRacesWithParticipantCountRequest());
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        RacesWithParticipantsResponse racesResponse = (RacesWithParticipantsResponse) response;
        return racesResponse.getRacesDTO();
    }

    @Override
    public void saveRaceEntries(List<RaceEntry> newEntities) throws ContestDataException {
        sendRequest(new CreateRaceEntriesRequest(newEntities));
    }

    @Override
    public Collection<Race> getRacesWhereNotRegisteredAndEngineCapacity(Long participantID, Integer engineCapacity) throws ContestDataException {
        sendRequest(new GetUnregisteredRacesRequest(participantID, engineCapacity));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        RacesByCapacityResponse racesResponse = (RacesByCapacityResponse) response;
        return racesResponse.getRaces();
    }

    @Override
    public Optional<Team> getTeamByName(String teamName) throws ContestDataException {
        sendRequest(new GetTeamByNameRequest(teamName));
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        TeamResponse teamResponse = (TeamResponse) response;
        return Optional.of(teamResponse.getTeam());
    }

    @Override
    public void saveParticipant(Participant newEntity) throws ContestDataException {
        sendRequest(new CreateParticipantRequest(newEntity));
    }

    @Override
    public void saveRace(Race newEntity) throws ContestDataException {
        sendRequest(new CreateRaceRequest(newEntity));
    }

    @Override
    public Collection<Team> findAllTeams() throws ContestDataException {
        sendRequest(new GetTeamsRequest());
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        AllTeamsResponse teamsResponse = (AllTeamsResponse) response;
        return teamsResponse.getTeams();
    }

    @Override
    public Collection<Participant> findAllParticipants() throws ContestDataException {
        sendRequest(new GetParticipantsRequest());
        Response response = readResponse();
        if (response instanceof ErrorResponse) {
            throw new ContestDataException(((ErrorResponse) response).getErrorMessage());
        }
        AllParticipantsResponse participantsResponse = (AllParticipantsResponse) response;
        return participantsResponse.getParticipants();
    }

    private void initializeConnection() throws ContestDataException {
        try {
            connection = new Socket(host, port);
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            finished = false;
            startResponseReader();
        } catch (IOException ioException) {
            ioException.printStackTrace();
        }
    }

    private void closeConnection() {
        finished = true;
        try {
            input.close();
            output.close();
            connection.close();
            client = null;
        } catch (IOException ioException) {
            ioException.printStackTrace();
        }
    }

    private void startResponseReader() {
        Thread responseReader = new Thread(new ReaderThread());
        responseReader.start();
    }

    private void handleUpdate(UpdateResponse update) {
        try {
            if (update instanceof final NewParticipantResponse participantResponse) {
                client.participantAdded(participantResponse.getParticipant());
            }

            if (update instanceof final NewRaceResponse raceResponse) {
                client.raceAdded(raceResponse.getRaceDTO());
            }

            if (update instanceof final UpdatedRacesResponse racesResponse) {
                client.raceEntriesAdded(racesResponse.getRacesDTO());
            }
        } catch (ContestDataException e) {
            e.printStackTrace();
        }
    }

    private void sendRequest(Request request) throws ContestDataException {
        try {
            output.writeObject(request);
            output.flush();
        } catch (IOException ioException) {
            throw new ContestDataException("Error sending object: " + ioException);
        }
    }

    private Response readResponse() throws ContestDataException {
        Response response;
        try {
            response = responseQueue.take();
        } catch (InterruptedException interruptedException) {
            throw new ContestDataException("Error reading response: " + interruptedException);
        }
        return response;
    }

    private class ReaderThread implements Runnable {
        public void run() {
            while (!finished) {
                try {
                    Object response = input.readObject();
                    System.out.println("response received " + response);
                    if (response instanceof UpdateResponse) {
                        handleUpdate((UpdateResponse) response);
                    } else {
                        try {
                            responseQueue.put((Response) response);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (IOException | ClassNotFoundException exception) {
                    System.out.println("reading error: " + exception);
                }
            }
        }
    }

}

File: Anul II/Medii de proiectare si programare/Laborator/lab08/C#/Ubb.BikeContest.Networking/Protocol/ServicesProxy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading.Tasks;
using Ubb.BikeContest.Model;
using Ubb.BikeContest.Model.data;
using Ubb.BikeContest.Networking.Protocol.Request;
using Ubb.BikeContest.Networking.Protocol.Response;
using Ubb.BikeContest.Services;
using Ubb.BikeContest.Services.src.Exceptions;

namespace Ubb.BikeContest.Networking.Protocol
{
    public class ServicesProxy : IContestServices
    {
        private readonly string host;
        private readonly int port;
        private IMainObserver client;
        private NetworkStream stream;
        private IFormatter formatter;
        private TcpClient connection;
        private readonly Queue<IResponse> responses;
        private volatile bool finished;
        private EventWaitHandle _waitHandle;

        public ServicesProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<IResponse>();
            InitialiseConnection();
        }

        public virtual User Login(string username, string passwordToken, IMainObserver client)
        {
            SendRequest(new LoginRequest(username, passwordToken));
            IResponse response = ReadResponse();
            if (response is OkResponse)
            {
                this.client = client;
                return ((OkResponse)response).User;
            }
            if (response is ErrorResponse)
            {
                ErrorResponse error = (ErrorResponse)response;
                CloseConnection();
                throw new ContestDataException(error.Message);
            }
            return null;
        }

        public void Logout(User user, IMainObserver client)
        {
            SendRequest(new LogoutRequest(user));
            IResponse response = ReadResponse();
            CloseConnection();
            if (response is ErrorResponse)
            {
                ErrorResponse error = (ErrorResponse)response;
                throw new ContestDataException(error.Message);
            }
        }

        private void CloseConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                _waitHandle.Close();
                client = null;
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception.StackTrace);
            }
        }

        private void SendRequest(IRequest request)
        {
            try
            {
                formatter.Serialize(stream, request);
                stream.Flush();
            }
            catch (Exception exception)
            {
                throw new ContestDataException("Error sending object " + exception);
            }
        }

        private IResponse ReadResponse()
        {
            IResponse response = null;
            try
            {
                _waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();
                }
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception.StackTrace);
            }
            return response;
        }

        private void InitialiseConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                _waitHandle = new AutoResetEvent(false);
                StartReader();
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception.StackTrace);
            }
        }

        private void StartReader()
        {
            Thread threadWorker = new Thread(Run);
            threadWorker.Start();
        }

        private void HandleUpdate(UpdateResponse update)
        {
            if (update is NewParticipantResponse)
            {
                NewParticipantResponse response = (NewParticipantResponse)update;
                Console.WriteLine("New participant added: " + response.Participant);
                try
                {
                    client.ParticipantAdded(response.Participant);
                }
                catch (ContestDataException exception)
                {
                    Console.WriteLine(exception.StackTrace);
                }
            }

            if (update is UpdatedRacesResponse)
            {
                UpdatedRacesResponse response = (UpdatedRacesResponse)update;
                Console.WriteLine("Updated races");
                try
                {
                    client.RaceEntriesAdded(response.Races);
                }
                catch (ContestDataException exception)
                {
                    Console.WriteLine(exception.StackTrace);
                }
            }
        }

        public virtual void Run()
        {
            while (!finished)
            {
                try
                {
                    object response = formatter.Deserialize(stream);
                    Console.WriteLine("Response received: " + response);
                    if (response is UpdateResponse)
                    {
                        HandleUpdate((UpdateResponse)response);
                    }
                    else
                    {
                        lock (responses)
                        {
                            responses.Enqueue((IResponse)response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (Exception exception)
                {
                    Console.WriteLine("Reading error: " + exception);
                }
            }
        }

        public List<RaceDto> GetRacesWithParticipantCount()
        {
            SendRequest(new GetRacesWithParticipantCountRequest());
            IResponse response = ReadResponse();
            if (response is ErrorResponse errorResponse)
            {
                throw new ContestDataException(errorResponse.Message);
            }
            RacesWithParticipantsResponse participantsResponse = (RacesWithParticipantsResponse)response;
            return participantsResponse.Races;
        }

        public List<Team> FindAllTeams()
        {
            SendRequest(new GetAllTeamsRequest());
            IResponse response = ReadResponse();
            if (response is ErrorResponse errorResponse)
            {
                throw new ContestDataException(errorResponse.Message);
            }
            AllTeamsResponse participantsResponse = (AllTeamsResponse)response;
            return participantsResponse.Teams;
        }

        public List<Participant> GetParticipantsByTeam(long id)
        {
            SendRequest(new GetParticipantsByTeamRequest(id));
            IResponse response = ReadResponse();
            if (response is ErrorResponse errorResponse)
            {
                throw new ContestDataException(errorResponse.Message);
            }
            GetParticipantsByTeamResponse participantsResponse = (GetParticipantsByTeamResponse)response;
            return participantsResponse.Participants;
        }

        public List<Participant> FindAllParticipants()
        {
            SendRequest(new GetAllParticipantsRequest());
            IResponse response = ReadResponse();
            if (response is ErrorResponse errorResponse)
            {
                throw new ContestDataException(errorResponse.Message);
            }
            AllParticipantsResponse participantsResponse = (AllParticipantsResponse)response;
            return participantsResponse.Participants;
        }

        public void SaveRaceEntries(List<RaceEntry> newEntities)
        {
            SendRequest(new CreateRaceEntriesRequest(newEntities));
        }

        public void SaveParticipant(Participant newEntity)
        {
            SendRequest(new CreateParticipantRequest(newEntity));
        }

        public List<Race> GetRacesWhereNotRegisteredAndEngineCapacity(long participantId, int engineCapacity)
        {
            SendRequest(new GetUnregisteredRacesRequest(participantId, engineCapacity));
            IResponse response = ReadResponse();
            if (response is ErrorResponse errorResponse)
            {
                throw new ContestDataException(errorResponse.Message);
            }
            UnregisteredRacesResponse racesResponse = (UnregisteredRacesResponse)response;
            return racesResponse.Races;
        }
    }
}
